packages/forge/src/governance/extensions/GovernorSorting.sol

```diff
@@ -2,232 +2,189 @@
-        internal
-        pure
-        returns (bool)
-    {
-        uint256 wMin = pos;
-        for (uint256 i = pos; i < netProposalVotes.length; i++) {
-            if (netProposalVotes[i] < netProposalVotes[wMin]) {
-                wMin = i;
+    // insert a new value into sortedRanks (this function is strictly O(n)).
+    // we know at this point it's:
+    //      -not tied
+    //      -the idx where it should go is in [0, smallestNonZeroSortedRanksValueIdx]
+    function _insertRank(
+        uint256 oldValue,
+        uint256 newValue,
+        uint256 smallestIdxMemVar,
+        uint256[] memory sortedRanksMemVar
+    ) internal {
+        // find the index to insert newValue at
+        uint256 insertingIndex;
+        for (uint256 index = 0; index < smallestIdxMemVar + 1; index++) {
+            if (newValue > sortedRanksMemVar[index]) {
+                insertingIndex = index;
+                break;
             }
         }
-        if (wMin == pos) return false;
-        int256 votesTmp = netProposalVotes[pos];
-        netProposalVotes[pos] = netProposalVotes[wMin];
-        netProposalVotes[wMin] = votesTmp;
-        uint256 proposalIdsTmp = proposalIds[pos];
-        proposalIds[pos] = proposalIds[wMin];
-        proposalIds[wMin] = proposalIdsTmp;
-        return true;
-    }
 
-    /**
-     * @dev Accessor to sorted list of proposalIds in ascending order.
-     */
-    function sortedProposals(bool excludeDeletedProposals)
-        public
-        view
-        virtual
-        returns (uint256[] memory sortedProposalIdsReturn)
-    {
-        (uint256[] memory proposalIdList, VoteCounts[] memory proposalVoteCountsArray) =
-            excludeDeletedProposals ? allProposalTotalVotesWithoutDeleted() : allProposalTotalVotes();
-        require(proposalIdList.length > 0, "GovernorSorting: cannot sort a list of zero length");
-        int256[] memory netProposalVotes = new int256[](proposalIdList.length);
-        for (uint256 i = 0; i < proposalVoteCountsArray.length; i++) {
-            netProposalVotes[i] = SafeCast.toInt256(proposalVoteCountsArray[i].forVotes)
-                - SafeCast.toInt256(proposalVoteCountsArray[i].againstVotes);
-        }
-        for (uint256 i = 0; i < proposalIdList.length - 1; i++) {
-            // Only goes to length minus 1 because sorting the last item would be redundant
-            _sortItem(i, netProposalVotes, proposalIdList);
+        // what we care about accounting for below is oldValue and the items at + below insertingIndex - we need to make sure all cases are accounted for:
+        //      - oldValue is at insertingIndex
+        //      - oldValue is at smallestIdxMemVar
+        //      - oldValue is between (exclusive) insertingIndex and smallestIdxMemVar (oldValue !> insertingIndex bc no downvoting)
+        //      - insertingIndex == smallestIdxMemVar
+        //      - smallestIdxMemVar + 1 == RANK_LIMIT
+        //      - smallestIdxMemVar + 1 < RANK_LIMIT
```

Comment:
Small contradiction here - prolly just the second one applicable?

Reply:
Yup, great call!